src/xwindows/keyboard.rs

```diff
@@ -0,0 +1,590 @@
+            }
+        };
+
+        Some((kc, self.get_key_modifiers()))
+    }
+
+    fn mod_is_active(&self, modifier: &str) -> bool {
+        // [TODO] consider state  Depressed & consumed mods
+        self.state
+            .borrow()
+            .mod_name_is_active(modifier, xkb::STATE_MODS_EFFECTIVE)
+    }
+
+    pub fn get_key_modifiers(&self) -> KeyModifiers {
+        let mut res = KeyModifiers::default();
+
+        if self.mod_is_active(xkb::MOD_NAME_SHIFT) {
+            res |= KeyModifiers::SHIFT;
+        }
+        if self.mod_is_active(xkb::MOD_NAME_CTRL) {
+            res |= KeyModifiers::CTRL;
+        }
+        if self.mod_is_active(xkb::MOD_NAME_ALT) {
+            // Mod1
+            res |= KeyModifiers::ALT;
+        }
+        if self.mod_is_active(xkb::MOD_NAME_LOGO) {
+            // Mod4
+            res |= KeyModifiers::SUPER;
+        }
+        if self.mod_is_active("Mod3") {
+            res |= KeyModifiers::SUPER;
+        }
+        //Mod2 is numlock
+        res
+    }
+
+    pub fn process_xkb_event(&self, connection: &xcb::Connection, event: &xcb::GenericEvent) {
+        let xkb_ev: &XkbGenericEvent = unsafe { xcb::cast_event(&event) };
+
+        if xkb_ev.device_id() == self.get_device_id() as u8 {
+            match xkb_ev.xkb_type() {
+                xcb::xkb::STATE_NOTIFY => {
+                    self.update_state(unsafe { xcb::cast_event(&event) });
+                }
+                xcb::xkb::MAP_NOTIFY | xcb::xkb::NEW_KEYBOARD_NOTIFY => {
+                    self.update_keymap(connection);
+                }
+                _ => {}
+            }
+        }
+    }
+    // for convenience, this fn takes &self, not &mut self
+    pub fn update_state(&self, ev: &xcb::xkb::StateNotifyEvent) {
+        self.state.borrow_mut().update_mask(
+            ev.base_mods() as xkb::ModMask,
+            ev.latched_mods() as xkb::ModMask,
+            ev.locked_mods() as xkb::ModMask,
+            ev.base_group() as xkb::LayoutIndex,
+            ev.latched_group() as xkb::LayoutIndex,
+            ev.locked_group() as xkb::LayoutIndex,
+        );
+    }
+
+    pub fn update_keymap(&self, connection: &xcb::Connection) {
+        let new_keymap = xkb::x11::keymap_new_from_device(
+            &self.context,
+            &connection,
+            self.get_device_id(),
+            xkb::KEYMAP_COMPILE_NO_FLAGS,
+        );
+        if new_keymap.get_raw_ptr() == std::ptr::null_mut() {
+            eprintln!("problem with new keymap");
+        }
+        let new_state = xkb::x11::state_new_from_device(&new_keymap, &connection, self.device_id);
+        if new_state.get_raw_ptr() == std::ptr::null_mut() {
+            eprintln!("problem with new state");
+        }
+        self.state.replace(new_state);
+        self.keymap.replace(new_keymap);
+        eprintln!("updated keymap");
+    }
+
+    pub fn get_device_id(&self) -> i32 {
+        self.device_id
+    }
+}
+
+#[cfg_attr(rustfmt, rustfmt_skip)]
```

Comment:
would it be the end of the world to let `rustfmt` run on this?
I value greatly the fact that a code formatter will mercilessly apply styling to code so that humans don't need to nitpick over the details in code reviews, even if it doesn't quite format to the taste of any one person :-)